[doc] Add CloudWatch integration documentation (#22638)
ray-project/ray: doc/source/cluster/aws-tips.rst

     region                <not set>             None    None

Please refer to `this discussion <https://github.com/ray-project/ray/issues/9327>`__ for more details.

.. _aws-cluster-cloudwatch:

Using Amazon CloudWatch
~~~~~~~~~~~~~~~~~~~~~~~

.. note::

>     Disclaimer: CloudWatch integration with Ray requires an AMI (or Docker image) with the Unified CloudWatch Agent pre-installed.

Reviewer: Agreed. Can we get a quick intro sentence on what CloudWatch is and why you need it? Then proceed with the above text, but without "note" or "disclaimer"

>     Disclaimer: CloudWatch integration with Ray requires an AMI (or Docker image) with the Unified CloudWatch Agent pre-installed.

Author: Added CloudWatch intro and removed the note.

    AMIs with the Unified CloudWatch Agent pre-installed are provided by the Amazon Ray Team, and are currently available in the us-east-1, us-east-2, us-west-1, and us-west-2 regions.
    Please direct any questions, comments, or issues to the Amazon Ray Team at `<https://github.com/amzn/amazon-ray/issues/new/choose>`_.

> .. list-table:: Unified CloudWatch Agent Images

Reviewer 2: Can we also provide an AWS CLI command to run to discover the latest image for a region? This would provide a nice fallback for customers to find the latest AMI prior to the updates being propagated back to this page.

    * - Base AMI
      - AMI ID
      - Region
      - Unified CloudWatch Agent Version

>     * - AWS Deep Learning AMI (Ubuntu 18.04, 64-bit)

Author: sure, added link to amzn-ray to AMI table title.

      - ami-069f2811478f86c20
      - us-east-1
      - v1.247348.0b251302
    * - AWS Deep Learning AMI (Ubuntu 18.04, 64-bit)
      - ami-058cc0932940c2b8b
      - us-east-2
      - v1.247348.0b251302
    * - AWS Deep Learning AMI (Ubuntu 18.04, 64-bit)
      - ami-044f95c9ef12883ef
      - us-west-1
      - v1.247348.0b251302
    * - AWS Deep Learning AMI (Ubuntu 18.04, 64-bit)
      - ami-0d88d9cbe28fac870
      - us-west-2
      - v1.247348.0b251302

.. note::

    Using Amazon CloudWatch will incur charges, please refer to `CloudWatch pricing <https://aws.amazon.com/cloudwatch/pricing/>`_ for details.

> Basic Setup:

Reviewer: Apart from the naming conventions, there should a) not be a colon (":") and b) you have to make the header one level deeper, i.e. this should be under CloudWatch in the side navigation. Same for the advanced setup.

> Basic Setup:

Author: Thanks Alex and Max, updated!

~~~~~~~~~~~~

> 1. Create a minimal YAML cluster config named ``cloudwatch-basic.yaml`` with the following contents:

Reviewer 3: i defer to you on naming convention here

> 1. Create a minimal YAML cluster config named ``cloudwatch-basic.yaml`` with the following contents:

Reviewer 2: Looks like the docs use "cluster config YAML" pretty consistently today, so my latest suggestions reflect this as the assumed standard.

> .. code-block:: yaml

Author: Thanks Alex, the ``cloudwatch-basic.yaml`` YAML file we provide here is a shorter version of what we have in [ray github repo](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-cloudwatch.yaml). 

If we want to let users just download it, we may need to open a PR to merge this shorter YAML file in ray repo as well .
May I know which option makes more sense, another ``cloudwatch-basic.yaml`` YAML file or just put code here?

            agent:
                # Path to Unified CloudWatch Agent config file
                config: "cloudwatch/example-cloudwatch-agent-config.json"
            dashboard:
                # CloudWatch Dashboard name
                name: "example-dashboard-name"
                # Path to the CloudWatch Dashboard config file
                config: "cloudwatch/example-cloudwatch-dashboard-config.json"

    auth:
        ssh_user: ubuntu

    available_node_types:
        ray.head.default:
            node_config:
            InstanceType: c5a.large
            ImageId: ami-0d88d9cbe28fac870  # Unified CloudWatch agent pre-installed AMI, us-west-2
            resources: {}
        ray.worker.default:
            node_config:
                InstanceType: c5a.large
                ImageId: ami-0d88d9cbe28fac870  # Unified CloudWatch agent pre-installed AMI, us-west-2
                IamInstanceProfile:
                    Name: ray-autoscaler-cloudwatch-v1
            resources: {}
            min_workers: 0

2. Download example CloudWatch agent and dashboard configuration files:

   First, create a ``cloudwatch`` directory in the same directory as ``cloudwatch-basic.yaml``.

>    Then, download the `CloudWatch agent config <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-agent-config.json>`_ and `CloudWatch dashboard config <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-dashboard-config.json>`_ to the cloudwatch directory.

Reviewer 3: Can we also provide some shell commands to download these?

Author: Sure, added command to download JSON config files.

3. Run ``ray up cloudwatch-basic.yaml`` to start your Ray Cluster.

.. note::

>     This will launch your Ray cluster in ``us-west-2`` by default. When launching a cluster for a different region, you'll need to change your YAML cluster launcher config file's ``region`` AND ``ImageId``.

Reviewer 3: This can just be part of a paragraph, doesn't need to be a note.

Author: Changed to a paragraph.

    See the ``Unified CloudWatch Agent Images table`` above for available AMIs by region.

4. Check out your Ray cluster's logs, metrics, and dashboard in the `CloudWatch Console <https://console.aws.amazon.com/cloudwatch/>`_!

Advanced Setup:
~~~~~~~~~~~~~~~

Refer to `example-cloudwatch.yaml <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-cloudwatch.yaml>`_ for a complete example.

> 1. Choose an AMI with the Unified CloudWatch Agent pre-installed.

Reviewer 3: Is it possible to set up your own AMI with the agent/are there aws docs/an explanation on how that would be done?

Reviewer 2: I think we could minimally provide a link to the CloudWatch agent installation docs (https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/install-CloudWatch-Agent-on-EC2-Instance.html) and the EC2 AMI creation docs (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html#creating-an-ami).

If we wanted to streamline the experience, we could also share a script that automates Ray CloudWatch Agent AMI creation. This actually used to be built into the AWS Cluster Launcher code for CloudWatch, but we eventually removed it since the install would sometimes take 30 minutes+ when run at EC2 instance launch time.

Author: Thanks Alex and Patrick, added instructions with link to AWS docs.

> 1. Choose an AMI with the Unified CloudWatch Agent pre-installed.

Reviewer 3: The levels of indentation here begin to feel a little extreme. Can we use sub headers instead of a list for this outermost one?

   Ensure that you're launching your Ray EC2 cluster in the same region as the AMI,
   then specify the ``ImageId`` to use with your cluster's head and worker nodes in your YAML cluster config file:

.. code-block:: yaml

    available_node_types:
        ray.head.default:
            node_config:
            InstanceType: c5a.large

>             ImageId: ami-0d88d9cbe28fac870 (us-west-2)

Reviewer 3: ```suggestion
            ImageId: ami-0d88d9cbe28fac870
```

        ray.worker.default:
            node_config:
            InstanceType: c5a.large

>             ImageId: ami-0d88d9cbe28fac870 (us-west-2)

Reviewer 3: ```suggestion
            ImageId: ami-0d88d9cbe28fac870
```

>    The AMIs are based on version 48 of the Ubuntu 18.04 AWS Deep Learning AMI, and ship with Unified CloudWatch Agent v1.247348.0b251302.

Reviewer 2: This is already noted in the above table and doesn't need to be restated here (the AMI version number is the only thing currently omitted from that table, but I also don't think it is critical to include initially).

    Extended EC2 metrics including CPU/Disk/Memory usage and process statistics can be found in the ``{cluster_name}-ray-CWAgent`` metric namespace.

    2. Dashboard: You will have a cluster-level dashboard showing total cluster CPUs and available object store memory.
    Process counts, disk usage, memory usage, and CPU utilization will be displayed as both cluster-level sums and single-node maximums/averages.

    3. Alarms: Node-level alarms tracking prolonged high memory, disk, and CPU usage are configured. Alarm actions are NOT set,
    and must be manually provided in your alarm config file.

   For more advanced CloudWatch configurations options, refer to `agent <https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-Configuration-File-Details.html>`_, `dashboard <https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/CloudWatch-Dashboard-Body-Structure.html>`_ and `alarm <https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricAlarm.html>`_ AWS documentations.

   CloudWatch Agent, Dashboard, and Alarm JSON config files support the following variables:

>     {instance_id}: Replaced with each EC2 instance ID in your Ray cluster.

Reviewer 3: ```suggestion
    ``{instance_id}``: Replaced with each EC2 instance ID in your Ray cluster.
```

Author: updated

>     {region}: Replaced with your Ray cluster's region.

Reviewer 3: ```suggestion
    ``{region}``: Replaced with your Ray cluster's region.
```

>     {cluster_name}: Replaced with your Ray cluster name.

Reviewer 3: ```suggestion
    ``{cluster_name}``: Replaced with your Ray cluster name.
```

    See CloudWatch Agent `Configuration File Details <https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-Configuration-File-Details.html>`_ for additional variables supported natively by the Unified CloudWatch Agent.

.. note::
    For CloudWatch alarm JSON configuration file, remember to add your alarm actions as shown below.

.. code-block:: json

     "AlarmActions":[
         "TODO: Add alarm actions! See https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html"
      ]

3. Reference your own CloudWatch JSON configuration files as file path in YAML cluster config file.

   You need to specify the file path to CloudWatch JSON configuration files relative to YAML cluster launcher config file like:

.. code-block:: yaml

>     agent:

Reviewer 3: should this be

```
provider:
  cloudwatch:
    agent:
      config: ...
```

Author: Yeah, nice catch! updated.

        config: "cloudwatch/example-cloudwatch-agent-config.json"

> 4. Setup minimal IAM permission required for CloudWatch.

Reviewer 3: What are the permissions? Can we list them out?

Reviewer 2: I think this is just bad wording here since all required IAM permissions will be granted via: https://github.com/amzn/amazon-ray/blob/main/python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py#L696-L730. We should just change this line to something like "Set your IAM Role and EC2 Instance Profile".

>    As for IAM role, ``ray-autoscaler-cloudwatch-v1`` is the default role for ensuring Unified CloudWatch Agent has permission to perform CloudWatch related operations.

Reviewer 3: Can we rewrite this as "by default ... if using your own instance profile, ensure it has ..."?

   A corresponding instance profile is created.
   Remember to use this default ``ray-autoscaler-cloudwatch-v1`` instance profile for worker nodes in YAML cluster config file to ensure ray worker nodes have the right permissions.

.. code-block:: yaml

    ray.worker.default:
        node_config:
            InstanceType: c5a.large
            IamInstanceProfile:
                Name: ray-autoscaler-cloudwatch-v1

5. To ship Ray system-level Prometheus metrics to CloudWatch.

   If you want to export Ray's Prometheus system metrics to CloudWatch, you should first ensure that your cluster has the
   Ray Dashboard installed, then uncomment the ``head_setup_commands`` section in `example-cloudwatch.yaml file <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-cloudwatch.yaml>`_ file.
   You can find Ray Prometheus metrics in the ``{cluster_name}-ray-prometheus`` metric namespace.

.. code-block:: yaml

>     head_setup_commands:

Reviewer 3: This is feeling pretty difficult to read/understand. Can we break it up into more commands, or use 

```
- bash -c "
mutli-line;
command;"
``` 

or something?

     region                <not set>             None    None

Please refer to `this discussion <https://github.com/ray-project/ray/issues/9327>`__ for more details.

.. _aws-cluster-cloudwatch:

Using Amazon CloudWatch
-----------------------

Amazon CloudWatch is a monitoring and observability service that provides data and actionable insights to monitor your applications, respond to system-wide performance changes, and optimize resource utilization.
CloudWatch integration with Ray requires an AMI (or Docker image) with the Unified CloudWatch Agent pre-installed.

AMIs with the Unified CloudWatch Agent pre-installed are provided by the Amazon Ray Team, and are currently available in the us-east-1, us-east-2, us-west-1, and us-west-2 regions.

> Please direct any questions, comments, or issues to the Amazon Ray Team at `amzn-ray <https://github.com/amzn/amazon-ray/issues/new/choose>`_.

Reviewer 2: ```suggestion
Please direct any questions, comments, or issues to the `Amazon Ray Team <https://github.com/amzn/amazon-ray/issues/new/choose>`_.
```

        availability_zone: us-west-2a
        # Start by defining a `cloudwatch` section to enable CloudWatch integration with your Ray cluster.
        cloudwatch:
            agent:
                # Path to Unified CloudWatch Agent config file
                config: "cloudwatch/example-cloudwatch-agent-config.json"
            dashboard:
                # CloudWatch Dashboard name
                name: "example-dashboard-name"
                # Path to the CloudWatch Dashboard config file
                config: "cloudwatch/example-cloudwatch-dashboard-config.json"

    auth:
        ssh_user: ubuntu

    available_node_types:
        ray.head.default:
            node_config:
            InstanceType: c5a.large
            ImageId: ami-0d88d9cbe28fac870  # Unified CloudWatch agent pre-installed AMI, us-west-2
            resources: {}
        ray.worker.default:
            node_config:
                InstanceType: c5a.large
                ImageId: ami-0d88d9cbe28fac870  # Unified CloudWatch agent pre-installed AMI, us-west-2
                IamInstanceProfile:
                    Name: ray-autoscaler-cloudwatch-v1
            resources: {}
            min_workers: 0

> 2. Download example CloudWatch agent and dashboard configuration files:

Reviewer 2: ```suggestion
2. Download example CloudWatch agent and dashboard configuration files.
```

First, create a ``cloudwatch`` directory in the same directory as ``cloudwatch-basic.yaml``.
Then, download the `CloudWatch agent config <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-agent-config.json>`_ and `CloudWatch dashboard config <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-dashboard-config.json>`_ to the cloudwatch directory.

.. code-block:: console

    $ mkdir cloudwatch
    $ cd cloudwatch
    $ wget https://raw.githubusercontent.com/ray-project/ray/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-agent-config.json
    $ wget https://raw.githubusercontent.com/ray-project/ray/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-dashboard-config.json

3. Run ``ray up cloudwatch-basic.yaml`` to start your Ray Cluster.
=================================================================================================

This will launch your Ray cluster in ``us-west-2`` by default. When launching a cluster for a different region, you'll need to change your cluster YAML config file's ``region`` AND ``ImageId``.

> See the ``Unified CloudWatch Agent Images table`` above for available AMIs by region.

Reviewer 2: ```suggestion
See the Unified CloudWatch Agent Images table above for available AMIs by region.
```

4. Check out your Ray cluster's logs, metrics, and dashboard in the `CloudWatch Console <https://console.aws.amazon.com/cloudwatch/>`_!
=======================================================================================================================================

> 

Reviewer 2: ```suggestion
A tail can be acquired on all logs written to a CloudWatch log group by ensuring that you have the `AWS CLI V2+ installed <https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html>`_ and then running:

.. code-block:: bash

aws logs tail $log_group_name --follow

```

3. Alarms: Node-level alarms tracking prolonged high memory, disk, and CPU usage are configured. Alarm actions are NOT set,
and must be manually provided in your alarm config file.

For more advanced CloudWatch configurations options, refer to `agent <https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-Configuration-File-Details.html>`_, `dashboard <https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/CloudWatch-Dashboard-Body-Structure.html>`_ and `alarm <https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricAlarm.html>`_ AWS documentations.

CloudWatch Agent, Dashboard, and Alarm JSON config files support the following variables:
``{instance_id}``: Replaced with each EC2 instance ID in your Ray cluster.

``{region}``: Replaced with your Ray cluster's region.

``{cluster_name}``: Replaced with your Ray cluster name.

See CloudWatch Agent `Configuration File Details <https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-Configuration-File-Details.html>`_ for additional variables supported natively by the Unified CloudWatch Agent.

.. note::

>     For CloudWatch alarm JSON configuration file, remember to add your alarm actions as shown below.

Reviewer 2: ```suggestion
    Remember to replace the ``AlarmActions`` placeholder in your CloudWatch Alarm config file!
```

.. code-block:: json

     "AlarmActions":[
         "TODO: Add alarm actions! See https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html"
      ]

3. Reference your own CloudWatch JSON configuration files as file path in cluster YAML config file.
===================================================================================================

> You need to specify the file path to CloudWatch JSON configuration files relative to cluster YAML config file like:

Reviewer 2: ```suggestion
You need to specify the file path to your CloudWatch JSON config files relative to your cluster config YAML like:
```

Also, is this correct? Does it actually need to be relative to the cluster config YAML, or relative to the working directory that, say, `ray up config.yaml` is run from?

Author: This should be relative to your cluster config YAML run from instead, my bad, thanks for pointing it out!

.. code-block:: yaml

     provider:
        cloudwatch:
            agent:
                config: "cloudwatch/example-cloudwatch-agent-config.json"

4. Set your IAM Role and EC2 Instance Profile.
==============================================

> By default, ``ray-autoscaler-cloudwatch-v1`` is the IAM role for ensuring Unified CloudWatch Agent has permission to perform CloudWatch related operations.

Reviewer 2: ```suggestion
By default the ``ray-autoscaler-cloudwatch-v1`` IAM role and EC2 instance profile is created at Ray cluster launch time. This role contains all additional permissions required to integrate CloudWatch with Ray.
```

A corresponding instance profile is created.

> Make sure to use this default ``ray-autoscaler-cloudwatch-v1`` instance profile for worker nodes in cluster YAML config file.

Reviewer 2: ```suggestion
You should ensure that all worker nodes are configured to use this default ``ray-autoscaler-cloudwatch-v1`` EC2 instance profile in your cluster config YAML.
```

> Additional managed policies attached to default role ray-autoscaler-cloudwatch-v1 are CloudWatchAgentAdminPolicy, AmazonSSMManagedInstanceCore, ssm:SendCommand, ssm:ListCommandInvocations, iam:PassRole.

Reviewer 2: ```suggestion
Managed policies attached to the default ``ray-autoscaler-cloudwatch-v1`` IAM role include ``CloudWatchAgentAdminPolicy``, ``AmazonSSMManagedInstanceCore``, ``ssm:SendCommand``, ``ssm:ListCommandInvocations``, and ``iam:PassRole``.
```

.. code-block:: yaml

    ray.worker.default:
        node_config:
            InstanceType: c5a.large
            IamInstanceProfile:
                Name: ray-autoscaler-cloudwatch-v1

5. To ship Ray system-level Prometheus metrics to CloudWatch.
=============================================================

> If you want to export Ray's Prometheus system metrics to CloudWatch, you should first ensure that your cluster has the

Reviewer 2: ```suggestion
To export Ray's Prometheus system metrics to CloudWatch, first ensure that your cluster has the
```

You can find Ray Prometheus metrics in the ``{cluster_name}-ray-prometheus`` metric namespace.

.. code-block:: yaml

    head_setup_commands:
    # Make `ray_prometheus_waiter.sh` executable.
    - RAY_INSTALL_DIR=`pip show ray | grep -Po "(?<=Location:).*"` && sudo chmod +x $RAY_INSTALL_DIR/ray/autoscaler/aws/cloudwatch/ray_prometheus_waiter.sh
    # Copy `prometheus.yml` to Unified CloudWatch Agent folder
    - RAY_INSTALL_DIR=`pip show ray | grep -Po "(?<=Location:).*"` && sudo cp -f $RAY_INSTALL_DIR/ray/autoscaler/aws/cloudwatch/prometheus.yml /opt/aws/amazon-cloudwatch-agent/etc
    # First get current cluster name, then let the Unified CloudWatch Agent restart and use `AmazonCloudWatch-ray_agent_config_{cluster_name}` parameter at SSM Parameter Store.
    - nohup sudo sh -c "`pip show ray | grep -Po "(?<=Location:).*"`/ray/autoscaler/aws/cloudwatch/ray_prometheus_waiter.sh `cat ~/ray_bootstrap_config.yaml | jq '.cluster_name'` >> '/opt/aws/amazon-cloudwatch-agent/logs/ray_prometheus_waiter.out' 2>> '/opt/aws/amazon-cloudwatch-agent/logs/ray_prometheus_waiter.err'" &

> 6. To update your Cloudwatch agent, dashboard and alarm JSON configuration files.

Reviewer 2: ```suggestion
6. Update CloudWatch Agent, Dashboard and Alarm config files.
```

=================================================================================

> If you need to change your CloudWatch agent/dashboard/alarm configurations, such as to include additional metrics,
> simply modify your local configs referenced in the Ray cluster YAML config file. Then run cli command ``ray up example-cloudwatch.yaml`` again.
> Unified Cloudwatch Agent will be restarted and additional metrics will be collected.

Reviewer 2: ```suggestion
You can apply changes to the CloudWatch Logs, Metrics, Dashboard, and Alarms for your cluster by simply modifying the CloudWatch config files referenced by your Ray cluster config YAML and re-running ``ray up example-cloudwatch.yaml``. The Unified CloudWatch Agent will be automatically restarted on all cluster nodes, and your config changes will be applied.
```

     region                <not set>             None    None

Please refer to `this discussion <https://github.com/ray-project/ray/issues/9327>`__ for more details.

.. _aws-cluster-cloudwatch:

Using Amazon CloudWatch
-----------------------

Amazon CloudWatch is a monitoring and observability service that provides data and actionable insights to monitor your applications, respond to system-wide performance changes, and optimize resource utilization.
CloudWatch integration with Ray requires an AMI (or Docker image) with the Unified CloudWatch Agent pre-installed.

AMIs with the Unified CloudWatch Agent pre-installed are provided by the Amazon Ray Team, and are currently available in the us-east-1, us-east-2, us-west-1, and us-west-2 regions.
Please direct any questions, comments, or issues to the `Amazon Ray Team <https://github.com/amzn/amazon-ray/issues/new/choose>`_.

> .. list-table:: `Unified CloudWatch Agent Images <https://github.com/amzn/amazon-ray>`_

Reviewer: In the rendered page this doesn't look like a table caption, but rather like a dangling link just sitting there. Could we maybe make this table caption read "All available unified CloudWatch agent images" and then have a full sentence linking to the `amazon-ray` repo after?

Reviewer 3: @Zyiqin-Miranda can you give me a ping after this and I'll merge?

Author: Thanks @maxpumperla for the review! Updated table name without the link.

Author: Sure, thanks @wuisawesome, updated and looks like passing ci tests

        availability_zone: us-west-2a
        # Start by defining a `cloudwatch` section to enable CloudWatch integration with your Ray cluster.
        cloudwatch:
            agent:
                # Path to Unified CloudWatch Agent config file
                config: "cloudwatch/example-cloudwatch-agent-config.json"
            dashboard:
                # CloudWatch Dashboard name
                name: "example-dashboard-name"
                # Path to the CloudWatch Dashboard config file
                config: "cloudwatch/example-cloudwatch-dashboard-config.json"

    auth:
        ssh_user: ubuntu

    available_node_types:
        ray.head.default:
            node_config:
            InstanceType: c5a.large
            ImageId: ami-0d88d9cbe28fac870  # Unified CloudWatch agent pre-installed AMI, us-west-2
            resources: {}
        ray.worker.default:
            node_config:
                InstanceType: c5a.large
                ImageId: ami-0d88d9cbe28fac870  # Unified CloudWatch agent pre-installed AMI, us-west-2
                IamInstanceProfile:
                    Name: ray-autoscaler-cloudwatch-v1
            resources: {}
            min_workers: 0

> 2. Download example CloudWatch agent and dashboard configuration files.

Reviewer 2: ```suggestion
2. Download CloudWatch Agent and Dashboard config.
```

Advanced Setup
~~~~~~~~~~~~~~

Refer to `example-cloudwatch.yaml <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/example-cloudwatch.yaml>`_ for a complete example.

1. Choose an AMI with the Unified CloudWatch Agent pre-installed.
=================================================================

Ensure that you're launching your Ray EC2 cluster in the same region as the AMI,
then specify the ``ImageId`` to use with your cluster's head and worker nodes in your cluster config YAML file.

The following CLI command returns the latest available Unified CloudWatch Agent Image for ``us-west-2``:

.. code-block:: bash

    aws ec2 describe-images --region us-west-2 --filters "Name=owner-id,Values=160082703681" "Name=name,Values=*cloudwatch*" --query 'Images[*].[ImageId,CreationDate]' --output text | sort -k2 -r | head -n1

.. code-block:: yaml

    available_node_types:
        ray.head.default:
            node_config:
            InstanceType: c5a.large
            ImageId: ami-0d88d9cbe28fac870
        ray.worker.default:
            node_config:
            InstanceType: c5a.large
            ImageId: ami-0d88d9cbe28fac870

> To build your own AMI with Unified CloudWatch Agent installed:

Reviewer 2: ```suggestion
To build your own AMI with the Unified CloudWatch Agent installed:
```

First, refer to `Installing the CloudWatch agent <https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/install-CloudWatch-Agent-on-EC2-Instance.html>`_ AWS docs to install Unified CloudWatch Agent on EC2 instance.
Then, refer to `creating-an-ami <https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/AMIs.html#creating-an-ami>`_ AWS docs to create an AMI from your EC2 instances.

2. Define your own CloudWatch agent, dashboard, alarm JSON configuration files.
===============================================================================

> You can start by using the example `agent config <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-agent-config.json>`_, `dashboard config <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-dashboard-config.json>`_ and `alarm config <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-alarm-config.json>`_ JSON configuration files.

Reviewer 2: ```suggestion
You can start by using the example CloudWatch `Agent <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-agent-config.json>`_, `Dashboard <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-dashboard-config.json>`_ and `Alarm <https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/aws/cloudwatch/example-cloudwatch-alarm-config.json>`_ config files.
```

> The example config files above provide the following default configurations:

Reviewer 2: ```suggestion
These example config files include the following features:
```

``{region}``: Replaced with your Ray cluster's region.

``{cluster_name}``: Replaced with your Ray cluster name.

See CloudWatch Agent `Configuration File Details <https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-Configuration-File-Details.html>`_ for additional variables supported natively by the Unified CloudWatch Agent.

.. note::
    Remember to replace the ``AlarmActions`` placeholder in your CloudWatch Alarm config file!

.. code-block:: json

     "AlarmActions":[
         "TODO: Add alarm actions! See https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html"
      ]

3. Reference your CloudWatch JSON config files in your cluster config YAML.
===========================================================================

You need to specify the file path to your CloudWatch JSON config files relative to where your cluster config YAML is run from like:

.. code-block:: yaml

     provider:
        cloudwatch:
            agent:
                config: "cloudwatch/example-cloudwatch-agent-config.json"

4. Set your IAM Role and EC2 Instance Profile.
==============================================

> By default the ``ray-autoscaler-cloudwatch-v1`` IAM role and EC2 instance profile is created at Ray cluster launch time. This role contains all additional permissions required to integrate CloudWatch with Ray.

Reviewer 2: ```suggestion
By default the ``ray-autoscaler-cloudwatch-v1`` IAM role and EC2 instance profile is created at Ray cluster launch time. This role contains all additional permissions required to integrate CloudWatch with Ray, namely the ``CloudWatchAgentAdminPolicy``, ``AmazonSSMManagedInstanceCore``, ``ssm:SendCommand``, ``ssm:ListCommandInvocations``, and ``iam:PassRole`` managed policies.
```

A corresponding instance profile is created.

You should ensure that all worker nodes are configured to use this default ``ray-autoscaler-cloudwatch-v1`` EC2 instance profile in your cluster config YAML.

> Managed policies attached to the default ``ray-autoscaler-cloudwatch-v1`` IAM role include ``CloudWatchAgentAdminPolicy``, ``AmazonSSMManagedInstanceCore``, ``ssm:SendCommand``, ``ssm:ListCommandInvocations``, and ``iam:PassRole``.

Reviewer 2: ```suggestion
```
(move to first paragraph as suggested above)
